### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Catonpancake**

### CODE REVIEW

File: dataloader.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+class DigitDataset(Dataset):
+    def __init__(self, img_dir, transform=None):
+        self.data = pd.read_csv(img_dir)
+        self.istrain = "label" in self.data.columns
+        self.img_dir = img_dir
+        self.img = self.data if not self.istrain else self.data.drop("label", axis=1)
+        self.img = [np.reshape(row[1].values,(28,28)) for row in self.img.iterrows()]
+        self.transform = transform
+        # self.target_transform = target_transform
```
Comment (1pha) [inline dataloader.py]: 안 쓰는 건 그때그떄 날려버리는 게 좋아요

File: dataloader.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        
+    def __len__(self):
+        return len(self.img)
+
+    def __getitem__(self,idx):
```
Comment (1pha) [inline dataloader.py]: Space (self, idx)
typehints: self에는 보통 안붙여줘요 (어차피 class라서). idx: int

File: dataloader.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            image = self.transform(image.astype(float))
+        if self.istrain:
+            label = self.data['label'][idx]
+            
+            return image, label
+        else:
+            return image
+
+        # if self.target_transform:
+        #     label = self.target_transform(label)
```
Comment (1pha) [inline dataloader.py]: 여기도 안쓰면 바로날려주세요

File: utils.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import os
+os.environ['KMP_DUPLICATE_LIB_OK']='True'
```
Comment (1pha) [inline utils.py:2]: https://github.com/NVlabs/stylegan3/issues/75
library 충돌이 나타나서 생기는 이슈로 보이는데, conda 가상환경을 하나 더 만들어서 정리하는 게 좋을 것 같아요.

File: dataloader.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import os
+os.environ['KMP_DUPLICATE_LIB_OK']='True'
```
Comment (1pha) [inline dataloader.py:2]: Library import의 경우는 나눠서 적어주는 것이 좋아요. 전에 몇 번 언급했지만 built-in / third-party / 직접개발한 모듈로 나눠서 쓰면 좋습니다.
이게 처음에 힘들면 isort를 사용해도 좋습니다 https://pycqa.github.io/isort/

File: utils.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import torchvision.transforms as transforms
+from sklearn.model_selection import train_test_split
+import timm
+import csv
+
+import seaborn as sns
+from PIL import Image
+import matplotlib.pyplot as plt
+
+def reshape(imgdata: pd.DataFrame()):
```
Comment (1pha) [inline utils.py]: typehint에는 class가 들어가야해요 (imgdata: pd.DataFrame) DataFrame 뒤에 () 지워주세요

File: utils.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    img_reshaped = np.reshape(imgdata.values,(28,28))
+
```
Comment (1pha) [inline utils.py:19]: enter remove

File: utils.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    return img_reshaped
+
+def drawimg(img, label=None):
+    if label == None:
+        plt.title(f'Label: {img[1]}')
+    else:
+        plt.title(f'Label: {label}')
```
Comment (1pha) [inline utils.py]: if문을 줄여줄 수 있어요
`plt.title(f"Label: {img[1] if label is None else label})`
혹은 더 간단하게
`plt.title(f"Label: {label or img[1]}")`
하지만 후자는 label과 img[1]에 뭐가 들어오느냐에 따라 잘 고려해야해요. or문은 둘 다 None이 아니고 유효한 값이 올 때 뒤의 값을 우선으로 사용합니다. (2 or 1은 1을 return, 1 or 2는 2를 return. 1 or None, None or 1은 모두 1을 반환하구요)

File: utils.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    plt.imshow(img[0], cmap='gray')
+
```
Comment (1pha) [inline utils.py:28]: 함수들 사이는 두 번 엔터를 쳐줍니다. 사소하지만 습관들여두면 좋아요!

File: utils.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+def multidrawimg(imgs, labels, batch_size: int=16):
+    fig = plt.figure(figsize=[10,10])
+    plt.subplots_adjust(top=0.9, wspace=0.2, hspace=0.35)
+    for i,data in enumerate(imgs):
+        ax = plt.subplot(4,batch_size//4,i+1)
+        ax.set_title(f'Label: {labels[i]}')
+        ax.imshow(tensor_to_pltimg(data), cmap='gray')
+        
+        
+def train(dataloader, model, loss_fn, optimizer, device):
```
Comment (1pha) [inline utils.py]: 전체적인 맥락에서 `train`과 `test`함수는 다른 스크립트로 분리되어 있는 게 좋을 것 같습니다.

Comment (1pha) [reply utils.py]: typehint도 적어주면 좋아요

File: utils.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            for X, y in dataloader:
+                X, y = X.to(device).float(), y.to(device)
+                pred = model(X)
+                test_loss += loss_fn(pred, y).item()
+                correct += (pred.argmax(1) == y).type(torch.float).sum().item()
+            test_loss /= num_batches
+            correct /= size
+            print(f"Test Error: \n Accuracy: {(100*correct):>0.1f}%, Avg loss: {test_loss:>8f} \n")
+        elif valid == False:
+            with open('Submission.csv', 'a+', newline='') as f:  # 'a+' means append to a file
```
Comment (1pha) [inline utils.py]: 이러면 같은 파일위에서 계속 작업하지 않나요? `test` 함수의 입력에서 fname을 따로 받아주거나, 같은 파일이 존재할 때 처리할 방법을 따로 적어주는 것이 좋을 것 같습니다.

**Review by 1pha — CHANGES_REQUESTED** [review]
제가 전에 봤을 때보다 정말 많이 깔끔해지신 것 같아요. 요청드린 사항들 반영해서 commit 넣어주시면 다시 봐드릴게요

File: utils.py (hunk 6 above)
Comment (Catonpancake) [reply utils.py]: Fixed

File: utils.py (hunk 7 above)
Comment (Catonpancake) [reply utils.py:19]: Fixed

File: utils.py (hunk 8 above)
Comment (Catonpancake) [reply utils.py]: 오 엄청 짧아지는군요

File: utils.py (hunk 9 above)
Comment (Catonpancake) [reply utils.py:28]: Fixed

File: utils.py (hunk 6 above)
Comment (1pha) [reply utils.py]: "Resolve conversation" 눌러주시면됩니당

File: utils.py (hunk 11 above)
Comment (Catonpancake) [reply utils.py]: 오 맞아요 이거 해두려고 했는데 고민되는게 버전번호를 넣는게 나을지 아니면 model 명이나 transform 변화같이 바뀌는 parameter들이 바뀌면 자연스럽게 제목이 바뀌게 하는 게 나을 지 고민중이었습니다. 보통은 어떻게 하는지 궁금하네요

File: utils_traintest.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
+from dataloader import DigitDataset
+from utils import reshape, drawimg, tensor_to_pltimg, multidrawimg, train, test
+import torch
+from torch import nn
+from torch.utils.data import Dataset
+from torch.utils.data import DataLoader
+import os
+os.environ['KMP_DUPLICATE_LIB_OK']='True'
+
+import csv
```
Comment (1pha) [inline utils_traintest.py]:
```suggestion
import os
import csv
os.environ['KMP_DUPLICATE_LIB_OK']='True'

import torch
from torch import nn
from torch.utils.data import Dataset
from torch.utils.data import DataLoader

from dataloader import DigitDataset
import utils as ut
```

그리고 뒤에 reshape이나 여러 utils에서 온 함수들 ut.reshape으로 사용하기
